Source/Core/VideoBackends/D3D12/D3DBase.cpp

```diff
@@ -0,0 +1,924 @@
+	default_root_signature->Release();
+
+	gpu_descriptor_heap_mgr->Release();
+	sampler_descriptor_heap_mgr->Release();
+	rtv_descriptor_heap_mgr->Release();
+	dsv_descriptor_heap_mgr->Release();
+
+	D3D::CleanupPersistentD3DTextureResources();
+
+	ULONG remaining_references = device12->Release();
+	if ((!s_debug_device12 && remaining_references) || (s_debug_device12 && remaining_references > 1))
+	{
+		ERROR_LOG(VIDEO, "Unreleased D3D12 references: %i.", remaining_references);
+	}
+	else
+	{
+		NOTICE_LOG(VIDEO, "Successfully released all D3D12 device references!");
+	}
+
+#if defined(_DEBUG) || defined(DEBUGFAST)
+	if (s_debug_device12)
+	{
+		--remaining_references; // the debug interface increases the refcount of the device, subtract that.
+		if (remaining_references)
+		{
+			// print out alive objects, but only if we actually have pending references
+			// note this will also print out internal live objects to the debug console
+			s_debug_device12->ReportLiveDeviceObjects(D3D12_RLDO_DETAIL);
+		}
+		SAFE_RELEASE(s_debug_device12);
+	}
+#endif
+
+	device12 = nullptr;
+	current_command_list = nullptr;
+
+	// unload DLLs
+	UnloadDXGI();
+	UnloadD3DCompiler();
+	UnloadD3D();
+}
+
+const std::string VertexShaderVersionString()
+{
+	return "vs_5_0";
+}
+
+const std::string GeometryShaderVersionString()
+{
+	return "gs_5_0";
+}
+
+const std::string PixelShaderVersionString()
+{
+	return "ps_5_0";
+}
+
+D3DTexture2D* &GetBackBuffer()
+{
+	return s_backbuf[s_current_back_buf];
+}
+
+unsigned int GetBackBufferWidth()
+{
+	return s_xres;
+}
+
+unsigned int GetBackBufferHeight()
+{
+	return s_yres;
+}
+
+// Returns the maximum width/height of a texture.
+unsigned int GetMaxTextureSize()
+{
+	return D3D11_REQ_TEXTURE2D_U_OR_V_DIMENSION;
+}
+
+void Reset()
+{
+	command_list_mgr->ExecuteQueuedWork(true);
+
+	// release all back buffer references
+	for (UINT i = 0; i < ARRAYSIZE(s_backbuf); i++)
+	{
+		SAFE_RELEASE(s_backbuf[i]);
+	}
+
+	D3D::command_list_mgr->ImmediatelyDestroyAllResourcesScheduledForDestruction();
+
+	// resize swapchain buffers
+	RECT client;
+	GetClientRect(hWnd, &client);
+	s_xres = client.right - client.left;
+	s_yres = client.bottom - client.top;
+
+	CheckHR(s_swapchain->ResizeBuffers(SWAP_CHAIN_BUFFER_COUNT, s_xres, s_yres, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
+
+	// recreate back buffer textures
+
+	HRESULT hr = S_OK;
+	ID3D12Resource* buf12 = nullptr;
+
+	for (UINT i = 0; i < SWAP_CHAIN_BUFFER_COUNT; i++)
+	{
+		ID3D12Resource* buf12;
+		hr = s_swapchain->GetBuffer(i, IID_PPV_ARGS(&buf12));
+
+		CHECK(SUCCEEDED(hr), "Create back buffer texture");
+
+		s_backbuf[i] = new D3DTexture2D(buf12,
+			D3D11_BIND_RENDER_TARGET,
+			DXGI_FORMAT_UNKNOWN,
+			DXGI_FORMAT_UNKNOWN,
+			DXGI_FORMAT_UNKNOWN,
+			false,
+			D3D12_RESOURCE_STATE_PRESENT
+			);
+
+		CHECK(s_backbuf != nullptr, "Create back buffer texture");
+
+		SAFE_RELEASE(buf12);
+		SetDebugObjectName12(s_backbuf[i]->GetTex12(), "backbuffer texture");
+	}
+
+	s_current_back_buf = 0;
```

Comment:
I did a quick check, and ResizeBackBuffers does always reset the index back to '0'.. do you think the API should still be called? Calling the API to me almost implies that it could be any value, when it will 'always' be zero.. but fine either way.